x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java

```diff
@@ -22,86 +23,62 @@
 import static org.elasticsearch.xpack.esql.expression.NamedExpressions.mergeOutputAttributes;
 
 /**
- * Looks up values from the {@code tables} request parameter.
- * <p>
- *     This class exists in two very different modes - an "unresolved" mode
- *     which contains only the {@link #tableName} and {@link #matchValues},
- *     and a "resolved" mode which contains everything. The process of
- *     resolving all of the references adds the extra columns by looking
- *     up the table from the analysis context.
- * </p>
+ * Looks up values from the associated {@code tables}.
+ * The class is used only during the analysis phased, after which it is replaced by a {@code Join}.
  */
 public class Lookup extends UnaryPlan {
-    private final Attribute tableName;
-    private final List<Attribute> matchFields;
-    private final List<TableColumnAttribute> matchValues;
-    private final List<TableColumnAttribute> mergeValues;
-    private List<Attribute> output;
-
-    public Lookup(
-        Source source,
-        LogicalPlan child,
-        Attribute tableName,
-        List<Attribute> matchFields,
-        @Nullable List<TableColumnAttribute> matchValues,
-        @Nullable List<TableColumnAttribute> mergeValues
-    ) {
+    private final Expression tableName;
+    private final List<NamedExpression> matchFields;
+    // initialized during the analysis phase for output and validation
+    // afterward, it is converted into a Join (BinaryPlan) hence why here it is not a child
+    private final LocalRelation localRelation;
+    private List<Attribute> lazyOutput;
+
+    public Lookup(Source source, LogicalPlan child, Expression tableName, List<NamedExpression> matchFields) {
+        this(source, child, tableName, matchFields, null);
+    }
+
+    public Lookup(Source source, LogicalPlan child, Expression tableName, List<NamedExpression> matchFields, LocalRelation localRelation) {
```

Comment:
The two ctors feels like overkill to me. We can pass `null` in during the tree walking.